bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java

```diff
@@ -0,0 +1,217 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs.record;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.RepresentationConfigurable;
+import org.bson.codecs.configuration.CodecConfigurationException;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.record.annotations.BsonProperty;
+import org.bson.codecs.record.annotations.BsonId;
+import org.bson.codecs.record.annotations.BsonRepresentation;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.RecordComponent;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.bson.assertions.Assertions.notNull;
+
+final class RecordCodec<T extends Record> implements Codec<T> {
+    private final Class<T> clazz;
+    private final Constructor<?> canonicalConstructor;
+    private final List<ComponentModel> componentModels;
+    private final ComponentModel componentModelForId;
+    private final Map<String, ComponentModel> fieldNameToComponentModel;
+
+    private static final class ComponentModel {
+        private final RecordComponent component;
+        private final Codec<?> codec;
+        private final int index;
+        private final String fieldName;
+
+        private ComponentModel(final RecordComponent component, final CodecRegistry codecRegistry, final int index) {
+            this.component = component;
+            this.codec = computeCodec(component, codecRegistry);
+            this.index = index;
+            this.fieldName = computeFieldName(component);
+        }
+
+        String getComponentName() {
+            return component.getName();
+        }
+
+        String getFieldName() {
+            return fieldName;
+        }
+
+        Object getValue(final Record record) throws InvocationTargetException, IllegalAccessException {
+            return component.getAccessor().invoke(record);
+        }
+
+        private static Codec<?> computeCodec(final RecordComponent component, final CodecRegistry codecRegistry) {
+            var codec = codecRegistry.get(toWrapper(component.getType()));
```

Comment:
Yes, the PojoCodec uses the `TypeData` which has a `boxType` method and a `isPrimitive` check

Reply:
I don't think it's worth refactoring to try to share this, but we can discuss again in final review

Reply:
If we move `TypeData` to an internal package and make `public` (it is currently package-access), then we can use `TypeData.builder(component.getType()).build().getType()` instead of `toWrapper(component.getType())`.